llvm/test/Transforms/InstCombine/pr71330.ll

```diff
@@ -0,0 +1,230 @@
+  %add11 = add nsw i64 %conv10, %inc24.us.lcssa24
+  %cmp12 = icmp slt i64 %add11, 15
+  br i1 %cmp12, label %for.body14, label %for.end25
+
+for.body14:                                       ; preds = %for.cond7
+  %inc = add i8 %inc.us.lcssa17, 1
+  br label %for.body20
+
+for.body20:                                       ; preds = %for.body20, %for.body14
+  %storemerge25 = phi i16 [ 0, %for.body14 ], [ %inc21, %for.body20 ]
+  %inc21 = add i16 %storemerge25, 1
+  %conv17 = sext i16 %inc21 to i32
+  %add18 = sub nsw i32 0, %conv17
+  %tobool19.not = icmp eq i32 %conv, %add18
+  br i1 %tobool19.not, label %for.cond15.for.inc23_crit_edge, label %for.body20
+
+for.cond15.for.inc23_crit_edge:                   ; preds = %for.body20
+  %split = phi i16 [ %inc21, %for.body20 ]
+  %inc24 = add nsw i64 %inc24.us.lcssa24, 1
+  br label %for.cond7
+
+for.end25:                                        ; preds = %for.cond7, %for.cond7.us.for.end25.split.us.split_crit_edge, %for.cond7.preheader.split.us.split, %for.cond7.us.us
+  %p3 = phi i16 [ %p1, %for.cond7.us.us ], [ 0, %for.cond7.us.for.end25.split.us.split_crit_edge ], [ %p0, %for.cond7.preheader.split.us.split ], [ %p2, %for.cond7 ]
+  %inc24.us.lcssa23 = phi i64 [ %inc24.us.lcssa27, %for.cond7.us.us ], [ %inc24.us.lcssa, %for.cond7.us.for.end25.split.us.split_crit_edge ], [ %inc24.us.lcssa2830, %for.cond7.preheader.split.us.split ], [ %inc24.us.lcssa24, %for.cond7 ]
+  %inc.us.lcssa16 = phi i8 [ %inc.us.lcssa20, %for.cond7.us.us ], [ %inc.us.lcssa, %for.cond7.us.for.end25.split.us.split_crit_edge ], [ %inc.us.lcssa2131, %for.cond7.preheader.split.us.split ], [ %inc.us.lcssa17, %for.cond7 ]
+  %tobool26.not = icmp eq i16 %p3, 0
+  br label %if.end
+
+if.end:                                           ; preds = %for.end25
+  %tobool28.not11 = icmp eq i8 %.pr31432, 0
+  br i1 %tobool28.not11, label %for.inc32, label %for.inc30.lr.ph
+
+for.inc30.lr.ph:                                  ; preds = %if.end
+  store i8 %la2, ptr @g, align 1
+  br i1 %tobool28.not, label %for.inc32, label %for.inc30.lr.ph.split
+
+for.inc30.lr.ph.split:                            ; preds = %for.inc30.lr.ph
+  %storemerge.lcssa12 = phi i32 [ %storemerge33, %for.inc30.lr.ph ]
+  br label %for.inc30
+
+for.inc30:                                        ; preds = %for.inc30, %for.inc30.lr.ph.split
+  br label %for.inc30
+
+for.inc32:                                        ; preds = %for.inc30.lr.ph, %if.end
+  %.pr313 = phi i8 [ %.pr31432, %if.end ], [ %la2, %for.inc30.lr.ph ]
+  %conv33 = trunc i32 %storemerge33 to i16
+  %add.i = add i16 %lb, %conv33
+  %conv35 = zext i16 %add.i to i32
+  %cmp = icmp slt i32 %conv35, 0
+  br i1 %cmp, label %for.cond7.preheader, label %for.end36
+
+for.end36:                                        ; preds = %for.inc32
+  %storemerge.lcssa = phi i32 [ %conv35, %for.inc32 ]
+  ret void
+
+; uselistorder directives
+  uselistorder i32 %storemerge33, { 2, 4, 1, 3, 0 }
+}
```

Comment:
I ended up reducing by hand, `llvm-reduce` simply doesn't change anything.

Reply:
What kind of interestingness test were you using? Diff between two opt binaries?

Reply:
I put an assertion in at the point of the bug and any change that triggered the assert.

Reply:
```
#! /usr/bin/env python3

import os
import sys
from common_util import run_proc


def os_do(cmd):
    print(cmd)
    return os.system(cmd)


s, r, stdout, stderr = run_proc(
    "./bin/opt -passes=instcombine {} -o unused.ll".format(sys.argv[1]),
    timeout=10,
    decode=True)
print("R={}".format(r))
if s is False:
    sys.exit(1)
if r != 134:
    sys.exit(1)
print(stderr)
if "UID:1234567890" in stderr:
    sys.exit(0)

sys.exit(1)

```

Reply:
Hm yeah, for some reason llvm-reduce is completely ineffective for this case. What worked for me is to manually remove uses of storemerge33 and then reorder them so the uselistorder is no longer necessary, and then run llvm-reduce afterwards. This gave me: https://gist.github.com/nikic/da8251d05a5df6038bd3e2eb773f64c0 This is probably still nowhere near minimal, but maybe good enough for a test case.

Reply:
Updated to your test case.